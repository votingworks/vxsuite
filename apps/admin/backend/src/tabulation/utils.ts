import { Tabulation } from '@votingworks/types';

/**
 * Tests whether CVR votes are empty.
 */
export function isBlankSheet(votes: Tabulation.Votes): boolean {
  return Object.values(votes).every((selections) => selections.length === 0);
}
