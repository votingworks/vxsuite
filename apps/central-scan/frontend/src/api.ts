import React from 'react';
import type { Api } from '@votingworks/central-scan-backend';
import {
  AUTH_STATUS_POLLING_INTERVAL_MS,
  QUERY_CLIENT_DEFAULT_OPTIONS,
} from '@votingworks/ui';
import {
  QueryClient,
  QueryKey,
  useMutation,
  useQuery,
  useQueryClient,
} from '@tanstack/react-query';
import * as grout from '@votingworks/grout';

export type ApiClient = grout.Client<Api>;

export function createApiClient(): ApiClient {
  return grout.createClient<Api>({ baseUrl: '/api' });
}

export const ApiClientContext = React.createContext<ApiClient | undefined>(
  undefined
);

export function useApiClient(): ApiClient {
  const apiClient = React.useContext(ApiClientContext);
  if (!apiClient) {
    throw new Error('ApiClientContext.Provider not found');
  }
  return apiClient;
}

export function createQueryClient(): QueryClient {
  return new QueryClient({ defaultOptions: QUERY_CLIENT_DEFAULT_OPTIONS });
}

// Queries

export const getAuthStatus = {
  queryKey(): QueryKey {
    return ['getAuthStatus'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getAuthStatus(), {
      refetchInterval: AUTH_STATUS_POLLING_INTERVAL_MS,
    });
  },
} as const;

export const getTestMode = {
  queryKey(): QueryKey {
    return ['getTestMode'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getTestMode());
  },
} as const;

export const getSystemSettings = {
  queryKey(): QueryKey {
    return ['getSystemSettings'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getSystemSettings());
  },
} as const;

export const getElectionDefinition = {
  queryKey(): QueryKey {
    return ['getElectionDefinition'];
  },
  useQuery() {
    const apiClient = useApiClient();
    return useQuery(this.queryKey(), () => apiClient.getElectionDefinition());
  },
} as const;

// Mutations

export const setTestMode = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.setTestMode, {
      async onSuccess() {
        await queryClient.invalidateQueries(getTestMode.queryKey());
      },
    });
  },
} as const;

export const checkPin = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.checkPin, {
      async onSuccess() {
        // Because we poll auth status with high frequency, this invalidation isn't strictly
        // necessary
        await queryClient.invalidateQueries(getAuthStatus.queryKey());
      },
    });
  },
} as const;

export const logOut = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.logOut, {
      async onSuccess() {
        // Because we poll auth status with high frequency, this invalidation isn't strictly
        // necessary
        await queryClient.invalidateQueries(getAuthStatus.queryKey());
      },
    });
  },
} as const;

export const updateSessionExpiry = {
  useMutation() {
    const apiClient = useApiClient();
    return useMutation(apiClient.updateSessionExpiry);
  },
} as const;

export const deleteBatch = {
  useMutation() {
    const apiClient = useApiClient();
    return useMutation(apiClient.deleteBatch);
  },
} as const;

export const configureFromBallotPackageOnUsbDrive = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.configureFromBallotPackageOnUsbDrive, {
      async onSuccess() {
        await queryClient.invalidateQueries(getSystemSettings.queryKey());
        await queryClient.invalidateQueries(getElectionDefinition.queryKey());
      },
    });
  },
} as const;

export const unconfigure = {
  useMutation() {
    const apiClient = useApiClient();
    const queryClient = useQueryClient();
    return useMutation(apiClient.unconfigure, {
      async onSuccess() {
        await queryClient.invalidateQueries(getTestMode.queryKey());
        await queryClient.invalidateQueries(getSystemSettings.queryKey());
        await queryClient.invalidateQueries(getElectionDefinition.queryKey());
      },
    });
  },
} as const;

export const clearBallotData = {
  useMutation() {
    const apiClient = useApiClient();
    return useMutation(apiClient.clearBallotData);
  },
} as const;
