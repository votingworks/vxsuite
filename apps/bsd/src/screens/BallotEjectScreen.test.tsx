import { waitFor, fireEvent, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import fetchMock from 'fetch-mock'
import React from 'react'
import { act } from 'react-dom/test-utils'
import {
  BallotType,
  AdjudicationReason,
  WriteInAdjudicationReasonInfo,
  UnmarkedWriteInAdjudicationReasonInfo,
} from '@votingworks/types'
import { typedAs } from '@votingworks/utils'
import { GetNextReviewSheetResponse } from '@votingworks/types/api/module-scan'
import BallotEjectScreen from './BallotEjectScreen'
import renderInAppContext from '../../test/renderInAppContext'

test('says the sheet is unreadable if it is', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: { type: 'BlankPage' },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: { type: 'BlankPage' },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { container, getByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })
})

test('says the ballot sheet is overvoted if it is', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: {
            type: 'InterpretedHmpbPage',
            markInfo: {
              ballotSize: { width: 1, height: 1 },
              marks: [],
            },
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 1,
            },
            adjudicationInfo: {
              requiresAdjudication: true,
              allReasonInfos: [
                {
                  type: AdjudicationReason.Overvote,
                  contestId: '1',
                  optionIds: ['1', '2'],
                  optionIndexes: [0, 1],
                  expected: 1,
                },
              ],
              enabledReasons: [AdjudicationReason.Overvote],
            },
            votes: {},
          },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: {
            type: 'InterpretedHmpbPage',
            markInfo: {
              ballotSize: { width: 1, height: 1 },
              marks: [],
            },
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 2,
            },
            adjudicationInfo: {
              requiresAdjudication: false,
              allReasonInfos: [],
              enabledReasons: [AdjudicationReason.Overvote],
            },
            votes: {},
          },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { container, getByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Original Ballot Removed'))
  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })

  continueScanning.mockClear()

  fireEvent.click(getByText('Tabulate Duplicate Ballot'))
  fireEvent.click(getByText('Tabulate Ballot and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({
    forceAccept: true,
    frontMarkAdjudications: [],
    backMarkAdjudications: [],
  })
})

test('says the ballot sheet is undervoted if it is', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: {
            type: 'InterpretedHmpbPage',
            markInfo: {
              ballotSize: { width: 1, height: 1 },
              marks: [],
            },
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 1,
            },
            adjudicationInfo: {
              requiresAdjudication: true,
              allReasonInfos: [
                {
                  type: AdjudicationReason.Undervote,
                  contestId: '1',
                  optionIds: [],
                  optionIndexes: [],
                  expected: 1,
                },
              ],
              enabledReasons: [AdjudicationReason.Undervote],
            },
            votes: {},
          },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: {
            type: 'InterpretedHmpbPage',
            markInfo: {
              ballotSize: { width: 1, height: 1 },
              marks: [],
            },
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 2,
            },
            adjudicationInfo: {
              requiresAdjudication: false,
              allReasonInfos: [],
              enabledReasons: [AdjudicationReason.Overvote],
            },
            votes: {},
          },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { container, getByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Original Ballot Removed'))
  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })

  continueScanning.mockClear()

  fireEvent.click(getByText('Tabulate Duplicate Ballot'))
  fireEvent.click(getByText('Tabulate Ballot and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({
    forceAccept: true,
    frontMarkAdjudications: [],
    backMarkAdjudications: [],
  })
})

test('says the ballot sheet is blank if it is', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: {
            type: 'InterpretedHmpbPage',
            markInfo: {
              ballotSize: { width: 1, height: 1 },
              marks: [],
            },
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 1,
            },
            adjudicationInfo: {
              requiresAdjudication: true,
              allReasonInfos: [{ type: AdjudicationReason.BlankBallot }],
              enabledReasons: [AdjudicationReason.Overvote],
            },
            votes: {},
          },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: {
            type: 'InterpretedHmpbPage',
            markInfo: {
              ballotSize: { width: 1, height: 1 },
              marks: [],
            },
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 2,
            },
            adjudicationInfo: {
              requiresAdjudication: false,
              allReasonInfos: [],
              enabledReasons: [AdjudicationReason.Overvote],
            },
            votes: {},
          },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { container, getByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Original Ballot Removed'))
  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })

  continueScanning.mockClear()

  fireEvent.click(getByText('Tabulate Duplicate Ballot'))
  fireEvent.click(getByText('Tabulate Ballot and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({
    forceAccept: true,
    frontMarkAdjudications: [],
    backMarkAdjudications: [],
  })
})

test('calls out live ballot sheets in test mode', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: {
            type: 'InvalidTestModePage',
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 1,
            },
          },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: {
            type: 'InvalidTestModePage',
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 2,
            },
          },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { container, getByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })
})

test('calls out test ballot sheets in live mode', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: {
            type: 'InvalidTestModePage',
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 1,
            },
          },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: {
            type: 'InvalidTestModePage',
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 2,
            },
          },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { container, getByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode={false} />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  expect(container).toMatchSnapshot()

  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })
})

test('shows invalid election screen when appropriate', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: {
            type: 'InvalidElectionHashPage',
            actualElectionHash: 'this-is-a-hash-hooray',
            expectedElectionHash: 'something',
          },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: { type: 'BlankPage' },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { getByText, queryAllByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode={false} />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  getByText('Wrong Election')
  getByText('Ballot Election Hash: this-is-a-')
  expect(queryAllByText('Tabulate Duplicate Ballot').length).toBe(0)

  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })
})

test('shows invalid election screen when appropriate', async () => {
  fetchMock.getOnce(
    '/scan/hmpb/review/next-sheet',
    typedAs<GetNextReviewSheetResponse>({
      interpreted: {
        id: 'mock-sheet-id',
        front: {
          image: { url: '/front/url' },
          interpretation: {
            type: 'InvalidPrecinctPage',
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 1,
            },
          },
        },
        back: {
          image: { url: '/back/url' },
          interpretation: {
            type: 'InvalidPrecinctPage',
            metadata: {
              ballotStyleId: '1',
              precinctId: '1',
              ballotType: BallotType.Standard,
              electionHash: '',
              isTestMode: false,
              locales: { primary: 'en-US' },
              pageNumber: 2,
            },
          },
        },
      },
      layouts: {},
      definitions: {},
    })
  )

  const continueScanning = jest.fn()

  const { getByText, queryAllByText } = renderInAppContext(
    <BallotEjectScreen continueScanning={continueScanning} isTestMode={false} />
  )

  await act(async () => {
    await waitFor(() => fetchMock.called)
  })

  getByText('Wrong Precinct')
  expect(queryAllByText('Tabulate Duplicate Ballot').length).toBe(0)

  fireEvent.click(getByText('Confirm Ballot Removed and Continue Scanning'))
  expect(continueScanning).toHaveBeenCalledWith({ forceAccept: false })
})

test('does NOT say ballot is blank if one side is blank and the other requires write-in adjudication (marked or unmarked)', async () => {
  for (const writeInReason of [
    AdjudicationReason.WriteIn,
    AdjudicationReason.UnmarkedWriteIn,
  ]) {
    fetchMock.getOnce(
      '/scan/hmpb/review/next-sheet',
      typedAs<GetNextReviewSheetResponse>({
        interpreted: {
          id: 'mock-sheet-id',
          front: {
            image: { url: '/writeinballot/front/url' },
            interpretation: {
              type: 'InterpretedHmpbPage',
              markInfo: {
                ballotSize: { width: 1, height: 1 },
                marks: [],
              },
              metadata: {
                ballotStyleId: '1',
                precinctId: '1',
                ballotType: BallotType.Standard,
                electionHash: '',
                isTestMode: false,
                locales: { primary: 'en-US' },
                pageNumber: 1,
              },
              adjudicationInfo: {
                requiresAdjudication: true,
                allReasonInfos: [{ type: AdjudicationReason.BlankBallot }],
                enabledReasons: [AdjudicationReason.BlankBallot, writeInReason],
              },
              votes: {},
            },
          },
          back: {
            image: { url: '/writeinballot/back/url' },
            interpretation: {
              type: 'InterpretedHmpbPage',
              markInfo: {
                ballotSize: { width: 1, height: 1 },
                marks: [],
              },
              metadata: {
                ballotStyleId: '1',
                precinctId: '1',
                ballotType: BallotType.Standard,
                electionHash: '',
                isTestMode: false,
                locales: { primary: 'en-US' },
                pageNumber: 2,
              },
              adjudicationInfo: {
                requiresAdjudication: true,
                allReasonInfos: [
                  {
                    type: writeInReason,
                    contestId: 'county-commissioners',
                    optionIndex: 0,
                    optionId: '__write-in-0',
                  } as
                    | WriteInAdjudicationReasonInfo
                    | UnmarkedWriteInAdjudicationReasonInfo,
                ],
                enabledReasons: [AdjudicationReason.BlankBallot, writeInReason],
              },
              votes: {},
            },
          },
        },
        layouts: {
          front: {
            ballotImage: {
              metadata: {
                ballotStyleId: '1',
                precinctId: '1',
                ballotType: BallotType.Standard,
                electionHash: '',
                isTestMode: false,
                locales: { primary: 'en-US' },
                pageNumber: 1,
              },
              imageData: {
                width: 20,
                height: 20,
              },
            },
            contests: [],
          },
          back: {
            ballotImage: {
              metadata: {
                ballotStyleId: '1',
                precinctId: '1',
                ballotType: BallotType.Standard,
                electionHash: '',
                isTestMode: false,
                locales: { primary: 'en-US' },
                pageNumber: 2,
              },
              imageData: {
                width: 20,
                height: 20,
              },
            },
            contests: [
              {
                bounds: { x: 0, y: 0, width: 10, height: 10 },
                corners: [
                  { x: 0, y: 0 },
                  { x: 1, y: 1 },
                  { x: 2, y: 2 },
                  { x: 3, y: 3 },
                ],
                options: [
                  {
                    bounds: { x: 0, y: 0, width: 10, height: 10 },
                    target: {
                      bounds: { x: 0, y: 0, width: 10, height: 10 },
                      inner: { x: 0, y: 0, width: 10, height: 10 },
                    },
                  },
                ],
              },
            ],
          },
        },
        definitions: {
          front: {
            contestIds: [],
          },
          back: {
            contestIds: ['county-commissioners'],
          },
        },
      })
    )

    const continueScanning = jest.fn()

    const { unmount } = renderInAppContext(
      <BallotEjectScreen continueScanning={continueScanning} isTestMode />
    )

    await act(async () => {
      await waitFor(() => fetchMock.called)
    })

    expect(screen.queryByText('Blank Ballot')).toBeNull()
    expect(screen.queryByText('Unknown Reason')).toBeNull()
    screen.getByText('Write-In')

    userEvent.type(
      screen.getByTestId(`write-in-input-__write-in-0`),
      'Lizard People'
    )

    // doubly nested acts() because there's setIsSaving(true) and then (false).
    await act(async () => {
      await act(async () => {
        userEvent.click(screen.getByText('Save & Continue Scanning'))
      })
    })

    expect(continueScanning).toHaveBeenCalledWith({
      forceAccept: true,
      frontMarkAdjudications: [],
      backMarkAdjudications: [
        {
          contestId: 'county-commissioners',
          isMarked: true,
          name: 'Lizard People',
          optionId: '__write-in-0',
          type: writeInReason,
        },
      ],
    })
    continueScanning.mockClear()

    unmount()
    fetchMock.reset()
  }
})
