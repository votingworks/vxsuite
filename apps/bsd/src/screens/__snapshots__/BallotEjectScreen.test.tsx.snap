// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls out live ballot sheets in test mode 1`] = `
<div>
  <div
    class="sc-dQppl CqIxD"
  >
    <div
      class="sc-iBPRYJ gplorY"
    >
      <div
        class="sc-fubCfw iqbrDX"
      >
        <div
          class="sc-pFZIQ McXrL"
        >
          <div
            class="sc-jrAGrp hkcHce"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-kEjbxe"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-iqHYGH eCkHsR"
        >
          <button
            class="sc-eCssSg czdiXU"
            type="button"
          >
            Confirm Ballot Removed and Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdfBwQ jA-DOsk"
    >
      <div
        class="sc-gsTCUz crfzwQ"
      >
        <div
          class="sc-kstrdz gkSOsj"
        >
          <div
            class="sc-dlfnbm ddzYNI"
          >
            <div
              class="sc-bqyKva dufLne"
            >
              Live Ballot
            </div>
            <p>
              This last scanned sheet 
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <p>
              Remove the LIVE ballot before continuing.
            </p>
          </div>
          <div
            class="sc-hBEYos IeiLU"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`calls out test ballot sheets in live mode 1`] = `
<div>
  <div
    class="sc-dQppl CqIxD"
  >
    <div
      class="sc-iBPRYJ gplorY"
    >
      <div
        class="sc-fubCfw iqbrDX"
      >
        <div
          class="sc-pFZIQ McXrL"
        >
          <div
            class="sc-jrAGrp hkcHce"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-kEjbxe"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-iqHYGH eCkHsR"
        >
          <button
            class="sc-eCssSg czdiXU"
            type="button"
          >
            Confirm Ballot Removed and Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdfBwQ jA-DOsk"
    >
      <div
        class="sc-gsTCUz crfzwQ"
      >
        <div
          class="sc-kstrdz gkSOsj"
        >
          <div
            class="sc-dlfnbm ddzYNI"
          >
            <div
              class="sc-bqyKva dufLne"
            >
              Test Ballot
            </div>
            <p>
              This last scanned sheet 
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <p>
              Remove the TEST ballot before continuing.
            </p>
          </div>
          <div
            class="sc-hBEYos IeiLU"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`says the ballot sheet is blank if it is 1`] = `
<div>
  <div
    class="sc-dQppl CqIxD"
  >
    <div
      class="sc-iBPRYJ gplorY"
    >
      <div
        class="sc-fubCfw iqbrDX"
      >
        <div
          class="sc-pFZIQ McXrL"
        >
          <div
            class="sc-jrAGrp hkcHce"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-kEjbxe"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-iqHYGH eCkHsR"
        >
          <button
            class="sc-eCssSg fEkUoC"
            disabled=""
            type="button"
          >
            Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdfBwQ jA-DOsk"
    >
      <div
        class="sc-gsTCUz crfzwQ"
      >
        <div
          class="sc-kstrdz gkSOsj"
        >
          <div
            class="sc-dlfnbm ddzYNI"
          >
            <div
              class="sc-bqyKva dufLne"
            >
              Unknown Reason
            </div>
            <p>
              This last scanned sheet 
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <h4>
              Original Ballot Scan
            </h4>
            <p>
              Remove ballot and create a duplicate ballot for the Resolution Board to review.
              <br />
              <button
                class="sc-eCssSg bFFZjC"
                type="button"
              >
                Original Ballot Removed
              </button>
            </p>
            <h4>
              Duplicate Ballot Scan
            </h4>
            <p>
              Confirm ballot sheet was reviewed by the Resolution Board and tabulate as ballow with issue which could not be determined.
              <br />
              <button
                class="sc-eCssSg bFFZjC"
                type="button"
              >
                Tabulate Duplicate Ballot
              </button>
            </p>
          </div>
          <div
            class="sc-hBEYos IeiLU"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`says the ballot sheet is overvoted if it is 1`] = `
<div>
  <div
    class="sc-dQppl CqIxD"
  >
    <div
      class="sc-iBPRYJ gplorY"
    >
      <div
        class="sc-fubCfw iqbrDX"
      >
        <div
          class="sc-pFZIQ McXrL"
        >
          <div
            class="sc-jrAGrp hkcHce"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-kEjbxe"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-iqHYGH eCkHsR"
        >
          <button
            class="sc-eCssSg fEkUoC"
            disabled=""
            type="button"
          >
            Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdfBwQ jA-DOsk"
    >
      <div
        class="sc-gsTCUz crfzwQ"
      >
        <div
          class="sc-kstrdz gkSOsj"
        >
          <div
            class="sc-dlfnbm ddzYNI"
          >
            <div
              class="sc-bqyKva dufLne"
            >
              Overvote
            </div>
            <p>
              This last scanned sheet 
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <h4>
              Original Ballot Scan
            </h4>
            <p>
              Remove ballot and create a duplicate ballot for the Resolution Board to review.
              <br />
              <button
                class="sc-eCssSg bFFZjC"
                type="button"
              >
                Original Ballot Removed
              </button>
            </p>
            <h4>
              Duplicate Ballot Scan
            </h4>
            <p>
              Confirm ballot sheet was reviewed by the Resolution Board and tabulate as ballot sheet with an
               
              <strong>
                overvote
              </strong>
              .
              <br />
              <button
                class="sc-eCssSg bFFZjC"
                type="button"
              >
                Tabulate Duplicate Ballot
              </button>
            </p>
          </div>
          <div
            class="sc-hBEYos IeiLU"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;

exports[`says the sheet is unreadable if it is 1`] = `
<div>
  <div
    class="sc-dQppl CqIxD"
  >
    <div
      class="sc-iBPRYJ gplorY"
    >
      <div
        class="sc-fubCfw iqbrDX"
      >
        <div
          class="sc-pFZIQ McXrL"
        >
          <div
            class="sc-jrAGrp hkcHce"
          >
            Voting
            <span>
              Works
            </span>
          </div>
          <div
            class="sc-kEjbxe"
          >
            Ballot Scanner
          </div>
        </div>
        <div
          class="sc-iqHYGH eCkHsR"
        >
          <button
            class="sc-eCssSg fEkUoC"
            disabled=""
            type="button"
          >
            Continue Scanning
          </button>
        </div>
      </div>
    </div>
    <main
      class="sc-bdfBwQ jA-DOsk"
    >
      <div
        class="sc-gsTCUz crfzwQ"
      >
        <div
          class="sc-kstrdz gkSOsj"
        >
          <div
            class="sc-dlfnbm ddzYNI"
          >
            <div
              class="sc-bqyKva dufLne"
            >
              Unreadable
            </div>
            <p>
              This last scanned sheet 
              <strong>
                was not tabulated
              </strong>
              .
            </p>
            <h4>
              Original Ballot Scan
            </h4>
            <p>
              Remove ballot and create a duplicate ballot for the Resolution Board to review.
              <br />
              <button
                class="sc-eCssSg bFFZjC"
                type="button"
              >
                Original Ballot Removed
              </button>
            </p>
            <h4>
              Duplicate Ballot Scan
            </h4>
            <p>
              Confirm sheet was reviewed by the Resolution Board and tabulate as 
              <strong>
                unreadable
              </strong>
              .
              <br />
              <button
                class="sc-eCssSg bFFZjC"
                type="button"
              >
                Tabulate Duplicate Ballot
              </button>
            </p>
          </div>
          <div
            class="sc-hBEYos IeiLU"
          >
            <div>
              <img
                alt="front"
                src="/front/url"
              />
            </div>
            <div>
              <img
                alt="back"
                src="/back/url"
              />
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
